EIPS/eip-1154.md

```diff
@@ -0,0 +1,94 @@
+---
+eip: 1154
+title: Oracle Interface
+author: Alan Lu (@cag)
+status: Draft
+type: Standards Track
+category: ERC
+created: 2018-06-13
+---
+
+## Simple Summary
+A standard interface for oracles.
+
+## Abstract
+In order for ethereum smart contracts to interact with off-chain systems, oracles must be used. These oracles report values which are normally off-chain, allowing smart contracts to react to the state of off-chain systems. A distinction and a choice is made between push and pull based oracle systems. Furthermore, a standard interface for oracles is described here, allowing different oracle implementations to be interchangeable.
+
+## Motivation
+The Ethereum ecosystem currently has many different oracle implementations available, but they do not provide a unified interface. Smart contract systems would be locked into a single set of oracle implementations, or they would require developers to write adapters/ports specific to the oracle system chosen in a given project.
+
+Beyond naming differences, there is also the issue of whether or not an oracle report-resolving transaction _pushes_ state changes by calling affected contracts, or changes the oracle state allowing dependent contracts to _pull_ the updated value from the oracle. These differing system semantics could introduce inefficiencies when adapting between them.
+
+Ultimately, the value in different oracle systems comes from their underlying resolution mechanics, and points where these systems are virtually identical should be standardized.
+
+## Specification
+
+<dl>
+  <dt>Oracle</dt>
+  <dd>An entity which reports data to the blockchain.</dd>
+
+  <dt>Oracle handler</dt>
+  <dd>A smart contract which receives data from an oracle.</dd>
+
+  <dt>ID</dt>
+  <dd>A way of indexing the data which an oracle reports. May be derived from or tied to a question for which the data provides the answer.</dd>
+
+  <dt>Result</dt>
+  <dd>Data associated with an id which is reported by an oracle. This data oftentimes will be the answer to a question tied to the id. Other equivalent terms that have been used include: answer, data, outcome.</dd>
+</dl>
+
+```solidity
+interface OracleHandler {
+    function receiveResult(bytes32 id, bytes32 result) external;
```

Comment:
This seems to assume one particular type of oracle - one that returns exactly 32 bytes of data, and is a trusted party. There are many other types of oracle; what about them?

Reply:
Regarding the trusted party factor: I've intentionally decided to start drafting the spec in as strict a manner as possible. With that said, there isn't a clear mandate about the authorization model, so it's not necessarily a single account which is authorized to make the report. Also, mechanisms like multisignature wallets, side/child chains, or something else may be used to distribute the trust if it was mandated to be a single account.

Regarding the 32 bytes of data: I am still debating and open to making the result an arbitrary-size blob of `bytes`. My contention is two-fold:

1. Putting much more than a word of result data (say like a paragraph) is probably not very actionable. Large amounts of input data is more suitable for, say, the field of machine learning, instead of smart contracts.
2. You can carve out an index byte in the ID if you want to pass 256 consecutive words to the handler, so it doesn't necessarily limit the result size in some sense, but at that point, I'd seriously ask whether this should even be going on the chain.

Reply:
Services like Oraclize would seem to demonstrate uses for more than 32 bytes of onchain data, however.

Reply:
Does anyone actually use more than 32 bytes of onchain data? I know that's how Oraclize *serves* the data as a string, but I reckon what contract authors are usually doing as soon as they get that data is to squidge whatever they get from there back into 32 bytes so they can actually use it...

We've assumed everything is a bytes32 for Reality Check; On our current scheme (implemented in our dapp, the contract only knows somebody sent it a bytes32) this is intended to map as follows:
 * 0 or 1 for binary
 * the number for an unsigned number
 * the number in twos-complement for a signed number
 * zero-based array for a single-choice question
 * one-based bitmask for multi-choice questions 

One hairy thing about this is that you often end up needing to express "this question is invalid" or "I couldn't answer this question". In Augur they call this "-1". (This is a slightly different thing to "have you reached a conclusion about this question" - which we handle separately). The natural thing is to encode it as 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff, but that clashes with an actual -1 as a signed number...